### PR TITLE
add logging features

### PR DESCRIPTION
**Pull request by wonderful89**
## add features

* it can remove the logs about `gc` and `flutter.frame` and it is redundant sometimes. 
image

* it can filter the logs by input. 
image

* 'real search' marks the timing to filter log.
* 'filter only summary' marks filter log by summary or details string.

That is what I need when developing.

### CODE REVIEW

**Comment by devoncarew** [comment]
Thanks for the PR! Can you add some screen shots of the feature, and describe it in a bit more detail? Thanks!

**Comment by wonderful89** [comment]
> Thanks for the PR! Can you add some screen shots of the feature, and describe it in a bit more detail? Thanks!

@devoncarew   I've updated it.

File: packages/devtools_app/lib/src/logging/html_logging_screen.dart [hunk 1]
```diff
@@ -2,9 +2,14 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
+import 'dart:async';
+
 import 'package:ansi_up/ansi_up.dart';
+import 'package:devtools_app/src/logging/checkout_box.dart';
```
Comment (jacob314) [inline packages/devtools_app/lib/src/logging/html_logging_screen.dart]: imports from within devtools_app should be relative imports.

File: packages/devtools_app/lib/src/logging/html_logging_screen.dart [hunk 2]
```diff
@@ -84,6 +119,31 @@ class HtmlLoggingScreen extends HtmlScreen {
...
+                ..clazz('margin-left')
+                ..click(() {
+                  final html.InputElement inputElement = textFiled.element;
+                  final String value = inputElement.value.trim();
+                  _search(value);
+                }),
+              PButton('Clear')
+                ..small()
+                ..click(() {
+                  print('test button click');
```
Comment (jacob314) [inline packages/devtools_app/lib/src/logging/html_logging_screen.dart]: remove all print statements from the cl.

File: packages/devtools_app/lib/src/logging/html_logging_screen.dart [hunk 3]
```diff
@@ -84,6 +119,31 @@ class HtmlLoggingScreen extends HtmlScreen {
...
+                  final html.InputElement inputElement = textFiled.element;
+                  inputElement.value = '';
+                  _search('');
+                }),
+              div()..flex(),
+              div()..flex(),
+              ToolBarCheckbox(ToolBarCheckboxDescription(name: 'hide some log', enabled: config[filterPlatformLogTag], tag: filterPlatformLogTag)).element,
```
Comment (jacob314) [inline packages/devtools_app/lib/src/logging/html_logging_screen.dart]: this checkbox needs a better name.
For example, maybe invert when it is checked and call it 'verbose'

File: packages/devtools_app/lib/src/logging/html_logging_screen.dart [hunk 4]
```diff
@@ -84,6 +119,31 @@ class HtmlLoggingScreen extends HtmlScreen {
...
+              div()..flex(),
+              ToolBarCheckbox(ToolBarCheckboxDescription(name: 'real search', enabled: config[realSearchTag], tag: realSearchTag)).element,
+              div()..flex(),
+              ToolBarCheckbox(ToolBarCheckboxDescription(name: 'filter only summary', enabled: config[filterSummaryTag], tag: filterSummaryTag)).element,
```
Comment (jacob314) [inline packages/devtools_app/lib/src/logging/html_logging_screen.dart]: what does this do?

File: packages/devtools_app/lib/src/logging/html_logging_screen.dart [hunk 5]
```diff
@@ -84,6 +119,31 @@ class HtmlLoggingScreen extends HtmlScreen {
...
+                  print('test button click');
+                  final html.InputElement inputElement = textFiled.element;
+                  inputElement.value = '';
+                  _search('');
+                }),
+              div()..flex(),
+              div()..flex(),
+              ToolBarCheckbox(ToolBarCheckboxDescription(name: 'hide some log', enabled: config[filterPlatformLogTag], tag: filterPlatformLogTag)).element,
+              div()..flex(),
+              ToolBarCheckbox(ToolBarCheckboxDescription(name: 'real search', enabled: config[realSearchTag], tag: realSearchTag)).element,
```
Comment (jacob314) [inline packages/devtools_app/lib/src/logging/html_logging_screen.dart]: call this 'filter' or 'search' rather than 'real search'

File: packages/devtools_app/lib/src/logging/html_logging_screen.dart [hunk 6]
```diff
@@ -125,10 +184,7 @@ class HtmlLoggingScreen extends HtmlScreen {
 
   HtmlTable<LogData> _createTableView() {
     final table = HtmlTable<LogData>.virtual();
-    table.model
-      ..addColumn(LogWhenColumn())
-      ..addColumn(LogKindColumn())
-      ..addColumn(LogMessageColumn(logMessageToHtml));
+    table.model..addColumn(LogWhenColumn())..addColumn(LogKindColumn())..addColumn(LogMessageColumn(logMessageToHtml));
```
Comment (jacob314) [inline packages/devtools_app/lib/src/logging/html_logging_screen.dart]: please run the dart formatter on the code.

File: packages/devtools_app/lib/src/logging/logging_controller.dart [hunk 7]
```diff
@@ -154,6 +160,11 @@ class LoggingController {
   }
 
   LoggingDetailsController detailsController;
+  static bool get hideSystemLog {
```
Comment (jacob314) [inline packages/devtools_app/lib/src/logging/logging_controller.dart]: why is this static?

File: packages/devtools_app/lib/src/logging/logging_controller.dart [hunk 8]
```diff
@@ -215,27 +227,36 @@ class LoggingController {
 
   void _updateStatus() {
     final int count = _loggingTableModel.rowCount;
-    final String label = count >= kMaxLogItemsLowerBound
-        ? '${nf.format(kMaxLogItemsLowerBound)}+'
-        : nf.format(count);
+    final String label = count >= kMaxLogItemsLowerBound ? '${nf.format(kMaxLogItemsLowerBound)}+' : nf.format(count);
```
Comment (jacob314) [inline packages/devtools_app/lib/src/logging/logging_controller.dart]: please run dartfmt with line length 80.

File: packages/devtools_app/lib/src/logging/logging_controller.dart [hunk 9]
```diff
@@ -215,27 +227,36 @@ class LoggingController {
...
     data.clear();
+    rawData.clear();
+    detailsController?.setData(null);
+    _loggingTableModel.setRows(data);
+  }
+
+  void search(String text) {
+    searchText = text;
+    data = rawData.where((item) {
+      if (!useSummaryFilter) {
```
Comment (jacob314) [inline packages/devtools_app/lib/src/logging/logging_controller.dart]: what is the use case for sometimes only searching the summary and sometimes searching the summary and details?

File: packages/devtools_app/lib/src/logging/logging_controller.dart [hunk 10]
```diff
@@ -249,18 +270,21 @@ class LoggingController {
...
+    if (hideSystemLog) {
+      return;
+    }
     // Events to show without a summary in the table.
     const Set<String> untitledEvents = {
       'Flutter.FirstFrame',
       'Flutter.FrameworkInitialization',
     };
 
+    print('_handleExtensionEvent: $e');
```
Comment (jacob314) [inline packages/devtools_app/lib/src/logging/logging_controller.dart]: remove all print statements.

File: packages/devtools_app/lib/src/logging/logging_controller.dart [hunk 11]
```diff
@@ -473,7 +492,19 @@ class LoggingController {
 
   void _log(LogData log) {
     // Insert the new item and clamped the list to kMaxLogItemsLength.
-    data.add(log);
+    if (searchText?.isNotEmpty ?? false) {
```
Comment (jacob314) [inline packages/devtools_app/lib/src/logging/logging_controller.dart]: I see this filtering logic in a couple of places. Add a helper method to determine whether a  LogData entry should be in data or rawData.

**Comment by wonderful89** [comment]
> Adding filtering to the log view is something we definitely need.
> Please reformat this cl with dartfmt at line length 80 and address the commends and then I'll re-review. right now there are a lot of diffs just due to whitespace changes so it is hard to review.
> This code also needs some tests to validate the filtering behavior.

Can you tell me how to run `flutter_drive` command? I try it ,but failed. So I can't add some tests.

```bash
(base) ➜  devtools_app git:(logging_feature_pr1) flutter drive --target=test/integration_tests/integration.dart  -d chrome
Test file not found: /Users/qqzhao/Work/Github/my/flutter_devtools/packages/devtools_app/test_driver/integration_tests/integration_test.dart
```

**Comment by jacob314** [comment]
https://github.com/flutter/devtools/blob/master/CONTRIBUTING.md#testing has instructions for running the tests. Our tests aren't yet typical Flutter integration tests.

Also you will need to sign the Google CLA to contribute code to our repo.
https://cla.developers.google.com/clas

**Comment by wonderful89** [comment]
> https://github.com/flutter/devtools/blob/master/CONTRIBUTING.md#testing has instructions for running the tests. Our tests aren't yet typical Flutter integration tests.
> 
> Also you will need to sign the Google CLA to contribute code to our repo.
> https://cla.developers.google.com/clas

I had updated the code  and add some tests.I had signed the Google CLA . Could you review it again ? Thank you.

File: packages/devtools_app/test/fixtures/logging_app.dart [hunk 12]
```diff
@@ -27,6 +27,13 @@ class Controller {
     log('emit log $count', name: 'logging');
   }
 
+  void emitLogTest() {
+    count++;
+
+    print('emitLog called test');
```
Comment (jacob314) [inline packages/devtools_app/test/fixtures/logging_app.dart]: please remove all print statements from this cl.

File: packages/devtools_app/test/integration_tests/logging.dart [hunk 13]
```diff
@@ -41,6 +41,35 @@ void loggingTests() {
...
+    final countAll = await logs.logCount();
+
+    await logs.filterLogs('test');
+    final countFilter = await logs.logCount();
+
+    await logs.filterLogs('');
+    final countAllAfter = await logs.logCount();
+
+    // Verify the log data shows up in the UI.
+    expect(countAll, greaterThanOrEqualTo(countFilter));
```
Comment (jacob314) [inline packages/devtools_app/test/integration_tests/logging.dart]: using greaterThanEqualTo doesn't verify that the filtering actually did anything
specify exactly how many log messages you expect in these two cases.

File: packages/devtools_app/lib/src/logging/checkout_box.dart [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+    ]);
+
+    _checkboxElement.checked = extensionDescription.enabled;
+    _checkboxElement.onChange.listen((_) {
+      final bool selected = _checkboxElement.checked;
+      config[extensionDescription.tag] = selected;
+    });
+  }
+
+  final ToolBarCheckboxDescription extensionDescription;
```
Comment (jacob314) [inline packages/devtools_app/lib/src/logging/checkout_box.dart]: this variable name is misleading.  This isn't anything to do with service extensions.

File: packages/devtools_app/lib/src/logging/checkout_box.dart [hunk 15]
```diff
@@ -0,0 +1,34 @@
+import 'package:html_shim/html.dart' as html;
+
+import '../../src/ui/html_elements.dart';
+import 'config.dart';
+
+class ToolBarCheckbox {
```
Comment (jacob314) [inline packages/devtools_app/lib/src/logging/checkout_box.dart]: Add the word Html to the class name as done elsewhere in the dart:html version of the app. We'll need to create a Flutter version of this class in the port of the app to flutter.

File: packages/devtools_app/lib/src/logging/html_logging_screen.dart [hunk 16]
```diff
@@ -84,6 +110,30 @@ class HtmlLoggingScreen extends HtmlScreen {
...
+                  _search(value);
+                }),
+              PButton('Clear')
+                ..small()
+                ..click(() {
+                  final html.InputElement inputElement = textFiled.element;
+                  inputElement.value = '';
+                  _search('');
+                }),
+              div()..flex(),
```
Comment (jacob314) [inline packages/devtools_app/lib/src/logging/html_logging_screen.dart]: why are there multiple flex elements in a row?

File: packages/devtools_app/lib/src/logging/html_logging_screen.dart [hunk 17]
```diff
@@ -70,6 +75,27 @@ class HtmlLoggingScreen extends HtmlScreen {
       true,
     );
 
+    final _search = (String text) {
+      controller.search(text);
+    };
+
+    final textFiled = CoreElement('input', classes: 'form-control input-sm')
```
Comment (jacob314) [inline packages/devtools_app/lib/src/logging/html_logging_screen.dart]: textFild --> searchInput

File: packages/devtools_app/lib/src/logging/html_logging_screen.dart [hunk 18]
```diff
@@ -70,6 +75,27 @@ class HtmlLoggingScreen extends HtmlScreen {
...
+      ..setAttribute('type', 'text')
+      ..setAttribute('style', 'width:480px;')
+      ..setAttribute('placeholder', 'input text')
+      ..id = 'uri-field';
```
Comment (jacob314) [inline packages/devtools_app/lib/src/logging/html_logging_screen.dart]: why is the id uri-field. That seems unrelated and generally we should not be adding ids.

File: packages/devtools_app/lib/src/logging/html_logging_screen.dart [hunk 19]
```diff
@@ -70,6 +75,27 @@ class HtmlLoggingScreen extends HtmlScreen {
...
+
+    textFiled.element.onKeyDown.listen((KeyboardEvent event) {
+      if (event.charCode != null) {
+//          event.preventDefault();
```
Comment (jacob314) [inline packages/devtools_app/lib/src/logging/html_logging_screen.dart]: remove commented out code.

File: packages/devtools_app/lib/src/logging/html_logging_screen.dart [hunk 20]
```diff
@@ -70,6 +75,27 @@ class HtmlLoggingScreen extends HtmlScreen {
...
+        Timer(const Duration(milliseconds: 20), () {
+          final html.InputElement inputElement = textFiled.element;
+          final String value = inputElement.value.trim();
+          _search(value);
+        });
+      }
+    });
+
```
Comment (jacob314) [inline packages/devtools_app/lib/src/logging/html_logging_screen.dart]: simplify this code to just
`_search(inputElement.value.trim());`

Comment (jacob314) [reply packages/devtools_app/lib/src/logging/html_logging_screen.dart]: rather than just setting a timer use the RateLimiter class. As is, you are still issuing the same # of requests as default but delaying the requests.

File: packages/devtools_app/lib/src/logging/html_logging_screen.dart [hunk 21]
```diff
@@ -84,6 +110,30 @@ class HtmlLoggingScreen extends HtmlScreen {
                   controller.clear();
                 }),
               div()..flex(),
+              textFiled,
+              PButton('Search')
+                ..small()
+                ..clazz('margin-left')
+                ..click(() {
+                  final html.InputElement inputElement = textFiled.element;
```
Comment (jacob314) [inline packages/devtools_app/lib/src/logging/html_logging_screen.dart]: this logic is repeated from earlier. Create a helper method.

File: packages/devtools_app/lib/src/logging/config.dart [hunk 22]
```diff
@@ -0,0 +1,5 @@
+const String platformLogTag = 'platformLog';
```
Comment (jacob314) [inline packages/devtools_app/lib/src/logging/config.dart]: rather than adding arbitrary string-value pairs to a Map, can you add this field directly to the controller?
You can use ValueListenable class to make the way the state is tracked easy to port to Flutter in the future.

**Comment by wonderful89** [comment]
> Looking close. Thanks for adding filtering support! That is a really important feature for a log view.

Thank you for your review. Your suggestions are very helpful to me . Anyway, I will suggest to add the logging filtering feature as soon as possible.

File: packages/devtools_app/lib/src/logging/html_logging_screen.dart (hunk 20 above)
Comment (wonderful89) [reply packages/devtools_app/lib/src/logging/html_logging_screen.dart]: It will lead to lose the last letter And I don't know why .

**Comment by jacob314** [comment]
lgtm
Once the tests are updated to pass we can merge.

**Comment by yk3372** [comment]
@jacob314 It's very useful and I need it very much. Please process this change as soon as possible, thank you

**Comment by yk3372** [comment]
@jacob314 
Could you help merge this change？

**Comment by devoncarew** [comment]
@yk3372 @wonderful89 - this PR has likely been superseded since it was initially created.  We're nearly complete with a re-write to the app to Flutter web. Additionally, the Flutter web version of the app now has a filter text field, so does contain some of the functionality that this PR would have added.

Can you file issues (one meta issue or separate issues) for the functionality you need from the logging view? We can ensure that the flutter web re-write already contains that functionality, or, look to add it in the near term if it looks like it would generalize to most users. Thanks!

**Comment by yk3372** [comment]
@devoncarew 
I am glad to receive your reply.

Now I'm using the Flutter on latest master branch and haven't seen any changes in the log in devtools. Below is the screenshot:
image

As a developer, logs is very important, but Flutter's log has a lot of unused info.
Especially when there is animation。

The following strong requirements, which can improve the development efficiency:
1. Filter the logs of the specified Kind, for example: flutter.frame、gc and so on.
2. Filter some required logs. For example, the terminal uses 'grep' to filter related logs. 
The terminal can use: flutter logs | grep 'xxx'

**Comment by devoncarew** [comment]
Great, thats for the input. I believe the new logging view supports the filtering / searching you mention.

The Flutter web version of this will be available to try out perhaps end of this month (you need to hit the lab button in the upper right of the toolbar). It'll likely be the default version the release after that.

**Comment by yk3372** [comment]
@devoncarew 
I am very glad to hear that.
But that Lab function is written by Flutter Web, and there was a problem with selection and copy text.

For example：
lab:
image
before:
image

I'm often use this log details to get JSON strings for parsing classes. However, the left and right entries in Lab cannot be selected and copy. 
Whether to consider using Selectable instead of Text, or other better methods, thanks

**Comment by devoncarew** [comment]
Good point; can you file an issue re: selectable text?

**Comment by yk3372** [comment]
> Good point; can you file an issue re: selectable text?

OK， I have made a new issue. Please optimization it，thanks
https://github.com/flutter/devtools/issues/1753

**Comment by yk3372** [comment]
> Great, thats for the input. I believe the new logging view supports the filtering / searching you mention.
> 
> The Flutter web version of this will be available to try out perhaps end of this month (you need to hit the lab button in the upper right of the toolbar). It'll likely be the default version the release after that.

@devoncarew I waited this for a month and didn't see this feature.
